source/extensions/filters/http/basic_auth/basic_auth_filter.cc

```diff
@@ -43,46 +42,49 @@ BasicAuthFilter::BasicAuthFilter(FilterConfigConstSharedPtr config) : config_(st
-
-    if (absl::StartsWith(auth_value, "Basic ")) {
-      // Extract and decode the Base64 part of the header.
-      absl::string_view base64Token = auth_value.substr(6);
-      const std::string decoded = Base64::decodeWithoutPadding(base64Token);
-
-      // The decoded string is in the format "username:password".
-      const size_t colon_pos = decoded.find(':');
-
-      if (colon_pos != std::string::npos) {
-        absl::string_view decoded_view = decoded;
-        absl::string_view username = decoded_view.substr(0, colon_pos);
-        absl::string_view password = decoded_view.substr(colon_pos + 1);
-
-        if (config_->validateUser(username, password)) {
-          config_->stats().allowed_.inc();
-          return Http::FilterHeadersStatus::Continue;
-        } else {
-          config_->stats().denied_.inc();
-          decoder_callbacks_->sendLocalReply(
-              Http::Code::Unauthorized,
-              "User authentication failed. Invalid username/password combination", nullptr,
-              absl::nullopt, "invalid_credential_for_basic_auth");
-          return Http::FilterHeadersStatus::StopIteration;
-        }
-      }
-    }
+
+  if (auth_header.empty()) {
+    return onDenied("User authentication failed. Missing username and password.",
+                    "no_credential_for_basic_auth");
   }
 
-  config_->stats().denied_.inc();
-  decoder_callbacks_->sendLocalReply(Http::Code::Unauthorized,
-                                     "User authentication failed. Missing username and password",
-                                     nullptr, absl::nullopt, "no_credential_for_basic_auth");
-  return Http::FilterHeadersStatus::StopIteration;
+  absl::string_view auth_value = auth_header[0]->value().getStringView();
+
+  if (!absl::StartsWith(auth_value, "Basic ")) {
+    return onDenied("User authentication failed. Expected 'Basic' authentication scheme.",
+                    "invalid_scheme_for_basic_auth");
+  }
+
+  // Extract and decode the Base64 part of the header.
+  absl::string_view base64_token = auth_value.substr(6);
+  const std::string decoded = Base64::decodeWithoutPadding(base64_token);
+
+  // The decoded string is in the format "username:password".
+  const size_t colon_pos = decoded.find(':');
+  if (colon_pos == std::string::npos) {
+    return onDenied("User authentication failed. Invalid basic credential format.",
+                    "invalid_format_for_basic_auth");
+  }
+
+  absl::string_view decoded_view = decoded;
+  absl::string_view username = decoded_view.substr(0, colon_pos);
+  absl::string_view password = decoded_view.substr(colon_pos + 1);
+
+  if (!config_->validateUser(username, password)) {
+    return onDenied("User authentication failed. Invalid username/password combination.",
+                    "invalid_credential_for_basic_auth");
+  }
+
+  config_->stats().allowed_.inc();
+  return Http::FilterHeadersStatus::Continue;
```

Comment:
This update treat `no auth header`, `has auth header but not for basic auth` and `has auth header but no colon` as three different error.